From: Michael Froman <mfroman@mozilla.com>
Date: Mon, 22 Dec 2025 20:19:23 -0600
Subject: (tmp-cherry-pick) Revert "Move CodecParameterMap functions to
 rtp_parameters" (3a563cebc1)

This reverts commit 882b287979da2d1a78e50849a3883bb1aad64a5c.

Reason for revert: Breaks downstream build during linking: backward reference detected: _ZN6webrtc27SessionDescriptionInterface6CreateENS_7SdpTypeENSt3__u10unique_ptrINS_18SessionDescriptionENS2_14default_deleteIS4_EEEENS2_17basic_string_viewIcNS2_11char_traitsIcEEEESB_NS2_6vectorINS_22IceCandidateCollectionENS2_9allocatorISD_EEEE in blaze-out/k8-opt/bin/third_party/webrtc/files/stable/webrtc/pc/_objs/webrtc_session_description_factory/webrtc_session_description_factory.pic.o refers to blaze-out/k8-opt/bin/third_party/webrtc/files/stable/webrtc/pc/_objs/webrtc_sdp/jsep_session_description.pic.o

I assume this is related to the circular dependency referenced in the comment of pc/BUILD.gn

Original change's description:
> Move CodecParameterMap functions to rtp_parameters
>
> This leaves webrtc_sdp.* with only the code related to parsing for
> SessionDescriptionInterface, unlocking the next step of co-locating that
> code with the jsep code which (circularly) depends on it.
>
> Bug: webrtc:42222470
> Change-Id: I460ea4f5fc47b5dafcefea2658a465f8ff97e84c
> Reviewed-on: https://webrtc-review.googlesource.com/c/src/+/414840
> Reviewed-by: Harald Alvestrand <hta@webrtc.org>
> Commit-Queue: Tomas Gunnarsson <tommi@webrtc.org>
> Cr-Commit-Position: refs/heads/main@{#45864}

Bug: webrtc:42222470
No-Presubmit: true
No-Tree-Checks: true
No-Try: true
Change-Id: I2dc766968073958cc2f30307e848d56a2bf97414
Reviewed-on: https://webrtc-review.googlesource.com/c/src/+/415341
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: Taylor Brandstetter <deadbeef@webrtc.org>
Reviewed-by: Tomas Gunnarsson <tommi@webrtc.org>
Reviewed-by: Mirko Bonadei <mbonadei@webrtc.org>
Commit-Queue: Tomas Gunnarsson <tommi@webrtc.org>
Cr-Commit-Position: refs/heads/main@{#45869}
---
 api/BUILD.gn                   |  3 --
 api/rtp_parameters.cc          | 79 --------------------------------
 api/rtp_parameters.h           | 13 ------
 api/rtp_parameters_unittest.cc | 19 --------
 pc/BUILD.gn                    | 13 +-----
 pc/rtc_stats_collector.cc      |  5 ++-
 pc/webrtc_sdp.cc               | 82 +++++++++++++++++++++++++++++++++-
 pc/webrtc_sdp.h                | 16 +++++++
 pc/webrtc_sdp_unittest.cc      | 21 +++++++++
 9 files changed, 121 insertions(+), 130 deletions(-)

diff --git a/api/BUILD.gn b/api/BUILD.gn
index 160a1f1d6e..a4fa721252 100644
--- a/api/BUILD.gn
+++ b/api/BUILD.gn
@@ -611,18 +611,15 @@ rtc_library("rtp_parameters") {
   deps = [
     ":array_view",
     ":priority",
-    ":rtc_error",
     ":rtp_transceiver_direction",
     "../media:media_constants",
     "../rtc_base:checks",
-    "../rtc_base:logging",
     "../rtc_base:stringutils",
     "../rtc_base/system:rtc_export",
     "video:resolution",
     "video_codecs:scalability_mode",
     "//third_party/abseil-cpp/absl/base:core_headers",
     "//third_party/abseil-cpp/absl/container:inlined_vector",
-    "//third_party/abseil-cpp/absl/strings",
     "//third_party/abseil-cpp/absl/strings:str_format",
     "//third_party/abseil-cpp/absl/strings:string_view",
   ]
diff --git a/api/rtp_parameters.cc b/api/rtp_parameters.cc
index 22d8e403c1..7bd443f711 100644
--- a/api/rtp_parameters.cc
+++ b/api/rtp_parameters.cc
@@ -15,56 +15,14 @@
 #include <tuple>
 #include <vector>
 
-#include "absl/strings/ascii.h"
 #include "absl/strings/string_view.h"
 #include "api/array_view.h"
-#include "api/rtc_error.h"
 #include "api/rtp_transceiver_direction.h"
 #include "media/base/media_constants.h"
 #include "rtc_base/checks.h"
-#include "rtc_base/logging.h"
-#include "rtc_base/string_encode.h"
 #include "rtc_base/strings/string_builder.h"
 
 namespace webrtc {
-namespace {
-constexpr char kSdpDelimiterSemicolon[] = ";";
-constexpr char kSdpDelimiterEqualChar = '=';
-constexpr char kSdpDelimiterEqual[] = "=";
-constexpr char kSdpDelimiterSemicolonChar = ';';
-
-void ParseFmtpParam(absl::string_view line,
-                    std::string* parameter,
-                    std::string* value) {
-  if (!tokenize_first(line, kSdpDelimiterEqualChar, parameter, value)) {
-    // Support for non-key-value lines like RFC 2198 or RFC 4733.
-    *parameter = "";
-    *value = std::string(line);
-  }
-  // a=fmtp:<payload_type> <param1>=<value1>; <param2>=<value2>; ...
-}
-
-bool IsFmtpParam(absl::string_view name) {
-  // RFC 4855, section 3 specifies the mapping of media format parameters to SDP
-  // parameters. Only ptime, maxptime, channels and rate are placed outside of
-  // the fmtp line. In WebRTC, channels and rate are already handled separately
-  // and thus not included in the CodecParameterMap.
-  return name != kCodecParamPTime && name != kCodecParamMaxPTime;
-}
-
-void WriteFmtpParameter(absl::string_view parameter_name,
-                        absl::string_view parameter_value,
-                        StringBuilder& os) {
-  if (parameter_name.empty()) {
-    // RFC 2198 and RFC 4733 don't use key-value pairs.
-    os << parameter_value;
-  } else {
-    // fmtp parameters: `parameter_name`=`parameter_value`
-    os << parameter_name << kSdpDelimiterEqual << parameter_value;
-  }
-}
-
-}  // namespace
 
 const char* DegradationPreferenceToString(
     DegradationPreference degradation_preference) {
@@ -83,43 +41,6 @@ const char* DegradationPreferenceToString(
 
 const double kDefaultBitratePriority = 1.0;
 
-bool WriteFmtpParameters(const CodecParameterMap& parameters,
-                         StringBuilder& os) {
-  bool empty = true;
-  const char* delimiter = "";  // No delimiter before first parameter.
-  for (const auto& entry : parameters) {
-    const std::string& key = entry.first;
-    const std::string& value = entry.second;
-
-    if (IsFmtpParam(key)) {
-      os << delimiter;
-      // A semicolon before each subsequent parameter.
-      delimiter = kSdpDelimiterSemicolon;
-      WriteFmtpParameter(key, value, os);
-      empty = false;
-    }
-  }
-
-  return !empty;
-}
-
-RTCError ParseFmtpParameterSet(absl::string_view line_params,
-                               CodecParameterMap& codec_params) {
-  // Parse out format specific parameters.
-  for (absl::string_view param :
-       split(line_params, kSdpDelimiterSemicolonChar)) {
-    std::string name;
-    std::string value;
-    ParseFmtpParam(absl::StripAsciiWhitespace(param), &name, &value);
-    if (codec_params.find(name) != codec_params.end()) {
-      RTC_LOG(LS_INFO) << "Overwriting duplicate fmtp parameter with key \""
-                       << name << "\".";
-    }
-    codec_params[name] = value;
-  }
-  return RTCError::OK();
-}
-
 RtcpFeedback::RtcpFeedback() = default;
 RtcpFeedback::RtcpFeedback(RtcpFeedbackType type) : type(type) {}
 RtcpFeedback::RtcpFeedback(RtcpFeedbackType type,
diff --git a/api/rtp_parameters.h b/api/rtp_parameters.h
index 9d6890c856..45e64eedc9 100644
--- a/api/rtp_parameters.h
+++ b/api/rtp_parameters.h
@@ -23,14 +23,12 @@
 #include "absl/strings/string_view.h"
 #include "api/media_types.h"
 #include "api/priority.h"
-#include "api/rtc_error.h"
 #include "api/rtp_transceiver_direction.h"
 #include "api/video/resolution.h"
 #include "api/video_codecs/scalability_mode.h"
 #include "rtc_base/system/rtc_export.h"
 
 namespace webrtc {
-class StringBuilder;
 
 using CodecParameterMap = std::map<std::string, std::string>;
 
@@ -139,17 +137,6 @@ RTC_EXPORT const char* DegradationPreferenceToString(
 
 RTC_EXPORT extern const double kDefaultBitratePriority;
 
-// Generates an FMTP line based on `parameters`. Please note that some
-// parameters are not considered to be part of the FMTP line, see the function
-// IsFmtpParam(). Returns true if the set of FMTP parameters is nonempty, false
-// otherwise.
-bool WriteFmtpParameters(const CodecParameterMap& parameters,
-                         StringBuilder& os);
-
-// Parses a string into an FMTP parameter set, in key-value format.
-RTCError ParseFmtpParameterSet(absl::string_view line_params,
-                               CodecParameterMap& codec_params);
-
 struct RTC_EXPORT RtcpFeedback {
   RtcpFeedbackType type = RtcpFeedbackType::CCM;
 
diff --git a/api/rtp_parameters_unittest.cc b/api/rtp_parameters_unittest.cc
index 0ce9a494d9..7d26e9d2b0 100644
--- a/api/rtp_parameters_unittest.cc
+++ b/api/rtp_parameters_unittest.cc
@@ -10,7 +10,6 @@
 
 #include "api/rtp_parameters.h"
 
-#include <string>
 #include <vector>
 
 #include "test/gtest.h"
@@ -302,22 +301,4 @@ TEST(RtpExtensionTest, FindHeaderExtensionByUri) {
                          extensions, kExtensionUri2,
                          RtpExtension::Filter::kRequireEncryptedExtension));
 }
-
-TEST(CodecParameterMap, ParsesKeyValueFmtpParameterSet) {
-  std::string params = "key1=value1;key2=value2";
-  CodecParameterMap codec_params;
-  ASSERT_TRUE(ParseFmtpParameterSet(params, codec_params).ok());
-  EXPECT_EQ(2U, codec_params.size());
-  EXPECT_EQ(codec_params["key1"], "value1");
-  EXPECT_EQ(codec_params["key2"], "value2");
-}
-
-TEST(CodecParameterMap, ParsesNonKeyValueFmtpParameterSet) {
-  std::string params = "not-in-key-value-format";
-  CodecParameterMap codec_params;
-  ASSERT_TRUE(ParseFmtpParameterSet(params, codec_params).ok());
-  EXPECT_EQ(1U, codec_params.size());
-  EXPECT_EQ(codec_params[""], "not-in-key-value-format");
-}
-
 }  // namespace webrtc
diff --git a/pc/BUILD.gn b/pc/BUILD.gn
index acd0895db1..721709e2ad 100644
--- a/pc/BUILD.gn
+++ b/pc/BUILD.gn
@@ -994,6 +994,7 @@ rtc_library("rtc_stats_collector") {
     ":sctp_data_channel",
     ":track_media_info_map",
     ":transport_stats",
+    ":webrtc_sdp",
     "../api:array_view",
     "../api:candidate",
     "../api:dtls_transport_interface",
@@ -1043,18 +1044,6 @@ rtc_library("rtc_stats_collector") {
     "//third_party/abseil-cpp/absl/functional:bind_front",
     "//third_party/abseil-cpp/absl/strings",
     "//third_party/abseil-cpp/absl/strings:string_view",
-
-    # TODO:bugs.webrtc.org/42222470 - This dependency shouldn't be here.
-    # `rtc_stats_collector` doesn't actually depend on this target. But due
-    # to a cyclical dependency between api/jsep.h and webrtc_sdp, this
-    # dependency has pulled in code from webrtc_sdp to other targets that
-    # depend on types declared in api/jsep.h and also have a dependency on
-    # `rtc_stats_collector`. The webrtc_sdp target depends on api/jsep.h, so
-    # that cycle prevents the dependencies from being set up correctly.
-    # Somehow (historically), this dependency here, has saved the day by
-    # pulling in the right symbols. Remove this once the remaining webrtc_sdp
-    # code gets co-located with jsep.h.
-    ":webrtc_sdp",
   ]
 }
 
diff --git a/pc/rtc_stats_collector.cc b/pc/rtc_stats_collector.cc
index 881638f236..8d6cc22426 100644
--- a/pc/rtc_stats_collector.cc
+++ b/pc/rtc_stats_collector.cc
@@ -64,6 +64,7 @@
 #include "pc/rtp_sender_proxy.h"
 #include "pc/rtp_transceiver.h"
 #include "pc/transport_stats.h"
+#include "pc/webrtc_sdp.h"
 #include "rtc_base/checks.h"
 #include "rtc_base/event.h"
 #include "rtc_base/ip_address.h"
@@ -105,7 +106,7 @@ std::string RTCCodecStatsIDFromTransportAndCodecParameters(
   // lines for the same PT and transport, which should be illegal SDP, then we
   // wouldn't need `fmtp` to be part of the ID here.
   StringBuilder fmtp;
-  if (WriteFmtpParameters(codec_params.parameters, fmtp)) {
+  if (WriteFmtpParameters(codec_params.parameters, &fmtp)) {
     sb << '_' << fmtp.Release();
   }
   return sb.str();
@@ -380,7 +381,7 @@ std::string GetCodecIdAndMaybeCreateCodecStats(
   }
 
   StringBuilder fmtp;
-  if (WriteFmtpParameters(codec_params.parameters, fmtp)) {
+  if (WriteFmtpParameters(codec_params.parameters, &fmtp)) {
     codec_stats->sdp_fmtp_line = fmtp.Release();
   }
   codec_stats->transport_id = transport_id;
diff --git a/pc/webrtc_sdp.cc b/pc/webrtc_sdp.cc
index 406543f307..f0f1954930 100644
--- a/pc/webrtc_sdp.cc
+++ b/pc/webrtc_sdp.cc
@@ -26,6 +26,7 @@
 
 #include "absl/algorithm/container.h"
 #include "absl/base/nullability.h"
+#include "absl/strings/ascii.h"
 #include "absl/strings/match.h"
 #include "absl/strings/str_cat.h"
 #include "absl/strings/string_view.h"
@@ -153,6 +154,8 @@ const char kSdpDelimiterSpace[] = " ";
 const char kSdpDelimiterSpaceChar = ' ';
 constexpr absl::string_view kSdpDelimiterColon = ":";
 const char kSdpDelimiterColonChar = ':';
+const char kSdpDelimiterSemicolon[] = ";";
+const char kSdpDelimiterSemicolonChar = ';';
 const char kSdpDelimiterSlashChar = '/';
 const char kNewLineChar = '\n';
 const char kReturnChar = '\r';
@@ -1008,12 +1011,24 @@ void WriteRtcpFbHeader(int payload_type, StringBuilder* os) {
   }
 }
 
+void WriteFmtpParameter(absl::string_view parameter_name,
+                        absl::string_view parameter_value,
+                        StringBuilder* os) {
+  if (parameter_name.empty()) {
+    // RFC 2198 and RFC 4733 don't use key-value pairs.
+    *os << parameter_value;
+  } else {
+    // fmtp parameters: `parameter_name`=`parameter_value`
+    *os << parameter_name << kSdpDelimiterEqual << parameter_value;
+  }
+}
+
 void AddFmtpLine(const Codec& codec, std::string* message) {
   StringBuilder os;
   WriteFmtpHeader(codec.id, &os);
   os << kSdpDelimiterSpace;
   // Create FMTP line and check that it's nonempty.
-  if (WriteFmtpParameters(codec.params, os)) {
+  if (WriteFmtpParameters(codec.params, &os)) {
     AddLine(os.str(), message);
   }
   return;
@@ -1418,6 +1433,14 @@ void BuildMediaDescription(const ContentInfo* content_info,
   }
 }
 
+bool IsFmtpParam(absl::string_view name) {
+  // RFC 4855, section 3 specifies the mapping of media format parameters to SDP
+  // parameters. Only ptime, maxptime, channels and rate are placed outside of
+  // the fmtp line. In WebRTC, channels and rate are already handled separately
+  // and thus not included in the CodecParameterMap.
+  return name != kCodecParamPTime && name != kCodecParamMaxPTime;
+}
+
 bool ParseConnectionData(absl::string_view line,
                          SocketAddress* addr,
                          SdpParseError* error) {
@@ -2014,6 +2037,20 @@ void UpdateCodec(MediaContentDescription* content_desc,
   AddOrReplaceCodec(content_desc, new_codec);
 }
 
+bool ParseFmtpParam(absl::string_view line,
+                    std::string* parameter,
+                    std::string* value,
+                    SdpParseError* error) {
+  if (!tokenize_first(line, kSdpDelimiterEqualChar, parameter, value)) {
+    // Support for non-key-value lines like RFC 2198 or RFC 4733.
+    *parameter = "";
+    *value = std::string(line);
+    return true;
+  }
+  // a=fmtp:<payload_type> <param1>=<value1>; <param2>=<value2>; ...
+  return true;
+}
+
 bool ParseFmtpAttributes(absl::string_view line,
                          const MediaType media_type,
                          MediaContentDescription* media_desc,
@@ -2049,7 +2086,7 @@ bool ParseFmtpAttributes(absl::string_view line,
 
   // Parse out format specific parameters.
   CodecParameterMap codec_params;
-  if (!ParseFmtpParameterSet(line_params, codec_params).ok()) {
+  if (!ParseFmtpParameterSet(line_params, codec_params, error)) {
     return false;
   }
 
@@ -3251,4 +3288,45 @@ std::unique_ptr<SessionDescriptionInterface> SdpDeserialize(
   return description;
 }
 
+bool WriteFmtpParameters(const CodecParameterMap& parameters,
+                         StringBuilder* os) {
+  bool empty = true;
+  const char* delimiter = "";  // No delimiter before first parameter.
+  for (const auto& entry : parameters) {
+    const std::string& key = entry.first;
+    const std::string& value = entry.second;
+
+    if (IsFmtpParam(key)) {
+      *os << delimiter;
+      // A semicolon before each subsequent parameter.
+      delimiter = kSdpDelimiterSemicolon;
+      WriteFmtpParameter(key, value, os);
+      empty = false;
+    }
+  }
+
+  return !empty;
+}
+
+bool ParseFmtpParameterSet(absl::string_view line_params,
+                           CodecParameterMap& codec_params,
+                           SdpParseError* error) {
+  // Parse out format specific parameters.
+  for (absl::string_view param :
+       split(line_params, kSdpDelimiterSemicolonChar)) {
+    std::string name;
+    std::string value;
+    if (!ParseFmtpParam(absl::StripAsciiWhitespace(param), &name, &value,
+                        error)) {
+      return false;
+    }
+    if (codec_params.find(name) != codec_params.end()) {
+      RTC_LOG(LS_INFO) << "Overwriting duplicate fmtp parameter with key \""
+                       << name << "\".";
+    }
+    codec_params[name] = value;
+  }
+  return true;
+}
+
 }  // namespace webrtc
diff --git a/pc/webrtc_sdp.h b/pc/webrtc_sdp.h
index fa4d722863..43a31e36f7 100644
--- a/pc/webrtc_sdp.h
+++ b/pc/webrtc_sdp.h
@@ -26,8 +26,12 @@
 #include "absl/base/nullability.h"
 #include "absl/strings/string_view.h"
 #include "api/jsep.h"
+#include "api/rtp_parameters.h"
+#include "rtc_base/strings/string_builder.h"
 
 namespace webrtc {
+struct SdpParseError;
+
 // Serializes the passed in SessionDescriptionInterface.
 // Serialize SessionDescription including candidates if
 // SessionDescriptionInterface has candidates.
@@ -45,6 +49,18 @@ absl_nullable std::unique_ptr<SessionDescriptionInterface> SdpDeserialize(
     absl::string_view sdp,
     SdpParseError* absl_nullable error = nullptr);
 
+// Generates an FMTP line based on `parameters`. Please note that some
+// parameters are not considered to be part of the FMTP line, see the function
+// IsFmtpParam(). Returns true if the set of FMTP parameters is nonempty, false
+// otherwise.
+bool WriteFmtpParameters(const CodecParameterMap& parameters,
+                         StringBuilder* os);
+
+// Parses a string into an FMTP parameter set, in key-value format.
+bool ParseFmtpParameterSet(absl::string_view line_params,
+                           CodecParameterMap& codec_params,
+                           SdpParseError* error);
+
 }  // namespace webrtc
 
 #endif  // PC_WEBRTC_SDP_H_
diff --git a/pc/webrtc_sdp_unittest.cc b/pc/webrtc_sdp_unittest.cc
index 39e4e6e97f..119fec6abe 100644
--- a/pc/webrtc_sdp_unittest.cc
+++ b/pc/webrtc_sdp_unittest.cc
@@ -4965,6 +4965,27 @@ TEST_F(WebRtcSdpTest, BackfillsDefaultFmtpValues) {
   EXPECT_EQ(value, "SRST");
 }
 
+TEST_F(WebRtcSdpTest, ParsesKeyValueFmtpParameterSet) {
+  std::string params = "key1=value1;key2=value2";
+  CodecParameterMap codec_params;
+  SdpParseError error;
+
+  ASSERT_TRUE(ParseFmtpParameterSet(params, codec_params, &error));
+  EXPECT_EQ(2U, codec_params.size());
+  EXPECT_EQ(codec_params["key1"], "value1");
+  EXPECT_EQ(codec_params["key2"], "value2");
+}
+
+TEST_F(WebRtcSdpTest, ParsesNonKeyValueFmtpParameterSet) {
+  std::string params = "not-in-key-value-format";
+  CodecParameterMap codec_params;
+  SdpParseError error;
+
+  ASSERT_TRUE(ParseFmtpParameterSet(params, codec_params, &error));
+  EXPECT_EQ(1U, codec_params.size());
+  EXPECT_EQ(codec_params[""], "not-in-key-value-format");
+}
+
 TEST_F(WebRtcSdpTest, SctpProtocolWithNonApplication) {
   std::string sdp =
       "v=0\r\n"
